File: app/services/agents/gatekeeper.py
import logging
import uuid
from app.schemas.agent_state import AgentState
from app.services.scraper import extract_product_info

logger = logging.getLogger(__name__)

def node_gatekeeper(state: AgentState) -> AgentState:
    """
    Validates inputs and determines routing.
    
    For chat-only flows (no image uploaded yet), we still allow the workflow
    to proceed to the Stylist — the Stylist can search for garments and give
    advice without requiring a user photo upfront.
    """
    logger.info("[Gatekeeper] Analyzing inputs")
    
    has_image = bool(state.get("user_image_url"))
    has_chat = bool(state.get("chat_history"))
    
    # Only reject if there's truly nothing to work with
    if not has_image and not has_chat:
        return {
            **state,
            "image_type": "invalid",
            "validation_message": "Please upload a photo or ask me something.",
            "current_agent": "gatekeeper"
        }
    
    # Check if a product URL was provided for scraping
    garment_page_url = state.get("garment_page_url")
    if garment_page_url and not state.get("garment_image_url"):
        logger.info("[Gatekeeper] External product URL detected. Initiating extraction...")
        try:
            extracted_data = extract_product_info(garment_page_url)
            if extracted_data:
                state["garment_image_url"] = extracted_data["image_url"]
                state["garment_title"] = extracted_data["title"]
                state["validation_message"] = f"Successfully extracted '{extracted_data['title']}'."
            else:
                state["validation_message"] = "Failed to extract product image from URL."
        except Exception as e:
            logger.warning(f"[Gatekeeper] Scraping failed: {e}")
            state["validation_message"] = "Could not extract product from URL."
    else:
        state["validation_message"] = "Input validated successfully."

    # Intercept Direct Try-on commands to populate UI state without LLM invocation
    chat_history = state.get("chat_history", [])
    if chat_history and chat_history[-1].get("content", "").startswith("I want to try on this item:"):
        latest_query = chat_history[-1]["content"]
        title = state.get("garment_title") or latest_query.replace("I want to try on this item: ", "").strip() or "Selected Item"
        state["styling_advice"] = f"Great choice! I'm sending the '{title}' to the fitting room right now."
        if state.get("garment_image_url"):
            state["recommended_items"] = [{
                "item_id": state.get("recommended_garment_id") or f"direct_{uuid.uuid4().hex[:8]}",
                "title": title,
                "image_url": state.get("garment_image_url")
            }]

    state["image_type"] = "full_body"
    state["current_agent"] = "gatekeeper"
    
    logger.info(f"[Gatekeeper] Passed — image={has_image}, chat={has_chat}")
    return state

File: src/lib/agents/gatekeeper.ts
// ─── Gatekeeper Agent ───────────────────────────────────────
// TypeScript port of DrapeNet's gatekeeper.py
// Validates inputs, scrapes product URLs, and routes requests.

import type { AgentState } from "./types";
import { extractProductInfo } from "../scraper";
import { randomUUID } from "crypto";

/**
 * Gatekeeper Agent — the pipeline's front door.
 * 
 * Responsibilities:
 * - Validates that there's something to work with (image or chat)
 * - Scrapes product URLs to extract garment images
 * - Intercepts direct "try on this item" commands for fast routing
 * - Sets image_type for downstream routing decisions
 */
export async function runGatekeeper(state: AgentState): Promise<AgentState> {
  console.log("[Gatekeeper] Analyzing inputs");

  const hasImage = Boolean(state.userImageUrl);
  const hasChat = state.chatHistory.length > 0;

  // Reject if there's nothing to work with
  if (!hasImage && !hasChat) {
    return {
      ...state,
      imageType: "invalid",
      validationMessage: "Please upload a photo or ask me something.",
      currentAgent: "gatekeeper",
    };
  }

  // If a product URL was provided, scrape it for the garment image
  if (state.garmentPageUrl && !state.garmentImageUrl) {
    console.log(
      "[Gatekeeper] External product URL detected. Initiating extraction..."
    );
    try {
      const extracted = await extractProductInfo(state.garmentPageUrl);
      if (extracted) {
        state.garmentImageUrl = extracted.imageUrl;
        state.garmentTitle = extracted.title;
        state.validationMessage = `Successfully extracted '${extracted.title}'.`;
      } else {
        state.validationMessage =
          "Failed to extract product image from URL.";
      }
    } catch (error) {
      console.warn(`[Gatekeeper] Scraping failed: ${error}`);
      state.validationMessage = "Could not extract product from URL.";
    }
  } else {
    state.validationMessage = "Input validated successfully.";
  }

  // Intercept direct try-on commands to skip Stylist LLM call
  const lastMessage = state.chatHistory[state.chatHistory.length - 1];
  if (lastMessage?.content?.startsWith("I want to try on this item:")) {
    const title =
      state.garmentTitle ||
      lastMessage.content.replace("I want to try on this item: ", "").trim() ||
      "Selected Item";

    state.stylingAdvice = `Great choice! I'm sending the '${title}' to the fitting room right now.`;
    state.intentType = "TRYON_SPECIFIC";

    if (state.garmentImageUrl) {
      state.recommendedItems = [
        {
          itemId:
            state.recommendedGarmentId ||
            `direct_${randomUUID().slice(0, 8)}`,
          title,
          imageUrl: state.garmentImageUrl,
        },
      ];
    }
  }

  state.imageType = "full_body";
  state.currentAgent = "gatekeeper";

  console.log(
    `[Gatekeeper] Passed — image=${hasImage}, chat=${hasChat}`
  );
  return state;
}

/**
 * Determines routing after Gatekeeper:
 * - "invalid" → end (bad input)
 * - Direct try-on command → artist (skip stylist)
 * - Chat history present → stylist
 * - Otherwise → artist
 */
export function routeGatekeeper(
  state: AgentState
): "stylist" | "artist" | "end" {
  if (state.imageType === "invalid") return "end";

  const lastMessage = state.chatHistory[state.chatHistory.length - 1];
  if (lastMessage?.content?.startsWith("I want to try on this item:")) {
    return "artist";
  }

  if (state.chatHistory.length > 0) {
    return "stylist";
  }

  return "artist";
}
